### DISCUSSION TITLE
driver sends rpc to firmware

### DISCUSSION THREAD

**Discussion by jeremyyx** [NVIDIA/open-gpu-kernel-modules#226] [General]
Is there a document summary all the API for sending the RPC interface to the firmware?

**Comment by mtijanic** [comment]
There is no documentation yet. The logical API for Kernel->Firmware calls is the same as User->Kernel ones, so https://github.com/NVIDIA/open-gpu-kernel-modules/discussions/157#discussioncomment-2756513 would be relevant here as well, but we need to write those docs first.

The RPC layer is not obfuscated, and if you want to go on an exploratory journey, you could breakpoint at e.g. [_kgspRpcSendMessage](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/src/nvidia/src/kernel/gpu/gsp/kernel_gsp.c#L109-L132) then look up the stack at which RPC type that is and what the params are.

**Reply by mtijanic** [reply]
Another way to go exploring is to just browse the [control calls](https://github.com/NVIDIA/open-gpu-kernel-modules/tree/main/src/common/sdk/nvidia/inc/ctrl). These account for probably 90% of the RPCs made to firmware and they're reasonably well documented.

**Reply by jeremyyx** [reply]
OK thanks!
